lib/puppet/util.rb

```diff
@@ -25,17 +25,47 @@ module Util
 
   # Run some code with a specific environment.  Resets the environment back to
   # what it was at the end of the code.
-  def self.withenv(hash)
-    saved = ENV.to_hash
-    hash.each do |name, val|
-      ENV[name.to_s] = val
-    end
+  # Windows can store unicode chars in the environment, so need to use API calls
+  #   to get and set the environment instead of ENV due to encoding problems.
+  def self.withenv(hash, mode=:posix)
```

Comment:
It might be a little cleaner (due to the nested `begin` / `ensure`) to leave the same basic control flow in place for this method, but to add a couple of different `private` helpers for posix vs windows, for saving and restoring the env vars.